### DISCUSSION TITLE
Is it possible to create a netbox script that exports a spreadsheet?

### DISCUSSION THREAD

**Discussion by CrawliestCotter** [netbox-community/netbox#17679] [Q&A]
Is it possible to create a netbox "Script" that exports a spreadsheet?

Specifically I'm looking to create an Event Rule that will run a script to export IP Addresses into a spreadsheet on the server whenever an IP Address is deletd, changed, or updated. 

I'm not sure what that python script would look like though. I've looked through the documentation on scripts, but it's not clear how I would access elements of an IP Address object, or tell the system to run an export to a specific folder.

**Comment by CrawliestCotter** [comment]
Also if there's an easier way to export schedule a spreadsheet export, I'd be happy to hear it.

**Comment by SaschaSchwarzK** [comment]
You can use an export template to export the data into a csv that you then can open with excel if that is what you want. 
You can also get the export via API get request as explained in the documentation.

GET /api/dcim/sites/?export=MyTemplateName

If you have an external system you could setup a web hook to trigger a script that call the API and saves the created file whre you want to have it.

The problem with a custom script is how you want to save the output? The option to return text, which can be a CSV format is not really suitable for a lot of data and you still would have to fetch the output data manually or via API. 
Or you include to upload it to some server, then you have to store credentials. Which you could store in an environment variable and access it from the script.
You could also install a python module to create excel files and use them in your custom script, but still you would need to upload them somewhere. 

Not sure why you want to export it to spreadsheets and not use the NetBox UI, but you might have your reasons.

**Reply by CrawliestCotter** [reply]
Thanks for the response!

I'm spearheading the implementation of netbox for my organization. I got my manager on board but he wants to make sure we have a csv export of certain parts of our database, incase we can't get to the netbox drive, or if he decides to move away from using it. We're starting small right now and are just importing data from spreadsheets we have with IPAM data, and that's the kind of data we're looking to export (IP's, Prefixes, and VLAN info)

The API request via a webhook sounds like a good way to go. I was looking to use the Event Rule system to make some sort of call to export to csv to a specific folder on our server whenever an edit is made.

Is the documentation you're talking about [here](https://netboxlabs.com/docs/netbox/en/stable/integrations/rest-api/#what-is-a-rest-api)

```have an external system you could setup a web hook to trigger a script that call the API and saves the created file whre you want to have it```
Would that be possible via the Event Rule system, so I could have it export whenever an edit is made to an ipam object?

**Reply by SaschaSchwarzK** [reply]
We use an Airflow server with Python DAGs for more complicated jobs. The Trigger is an event rule that uses a web hook to call the Airflow API to start the job. That python DAG then connects the Netbox API to collect the data and processes it. IF you just need the data for the specific object that was change, then you can include the pre and port data for that object in the payload of the Webhook.

**Comment by ross-cello** [comment]
We use `openpyxl` in our scripts to generate XLSX documents (also read as templates), added to local_requirements.txt and installed in NB's venv.

Not sure about "easy", but it's a method.
`pandas` may be an alternative and imo is cleaner coding... but I've personally not bothered with a migration.

CSV is much easier if you're okay with that format.

**Comment by llamafilm** [comment]
A custom script can write data to a Google Sheet. https://developers.google.com/sheets/api/quickstart/python

**Comment by pva** [comment]
https://netboxlabs.com/blog/how-to-generate-cabling-cut-sheets-with-netbox-export-templates/
